concent_api/core/models.py

```diff
@@ -73,3 +79,252 @@ class MessageAuth(Model):
+    assert set(PASSIVE_STATES) & set(ACTIVE_STATES) == set()
+
+    # Defines possible state transitions, where keys are transitions to, and values are lists of transitions from.
+    # None means transition from Unknown state.
+    POSSIBLE_TRANSITIONS_TO = {
+        SubtaskState.FORCING_REPORT: {
+            None,
+        },
+        SubtaskState.REPORTED: {
+            SubtaskState.FORCING_REPORT,
+        },
+        SubtaskState.FORCING_RESULT_TRANSFER: {
+            None,
+            SubtaskState.REPORTED,
+        },
+        SubtaskState.RESULT_UPLOADED: {
+            SubtaskState.FORCING_RESULT_TRANSFER,
+        },
+        SubtaskState.FORCING_ACCEPTANCE: {
+            None,
+            SubtaskState.REPORTED,
+            SubtaskState.RESULT_UPLOADED,
+        },
+        SubtaskState.REJECTED: {
+            SubtaskState.FORCING_ACCEPTANCE,
+        },
+        SubtaskState.VERIFICATION_FILE_TRANSFER: {
+            None,
+            SubtaskState.REPORTED,
+            SubtaskState.RESULT_UPLOADED,
+            SubtaskState.FORCING_ACCEPTANCE,
+            SubtaskState.REJECTED,
+        },
+        SubtaskState.ADDITIONAL_VERIFICATION: {
+            SubtaskState.VERIFICATION_FILE_TRANSFER,
+        },
+        SubtaskState.ACCEPTED: {
+            SubtaskState.FORCING_ACCEPTANCE,
+            SubtaskState.ADDITIONAL_VERIFICATION,
+        },
+        SubtaskState.FAILED: {
+            SubtaskState.FORCING_REPORT,
+            SubtaskState.FORCING_RESULT_TRANSFER,
+            SubtaskState.VERIFICATION_FILE_TRANSFER,
+            SubtaskState.ADDITIONAL_VERIFICATION,
+        },
+    }
+
+    assert set(POSSIBLE_TRANSITIONS_TO) == set(SubtaskState)
+    assert set(SubtaskState) | {None} == {
+        state
+        for from_states in POSSIBLE_TRANSITIONS_TO.values()
+        for state in from_states  # type: ignore
+    } | {SubtaskState.FAILED, SubtaskState.ACCEPTED}, 'There are no transitions from some states'
+
+    # Defines in which states given related message can't be None.
+    REQUIRED_RELATED_MESSAGES_IN_STATES = {
+        'task_to_compute': {
+            SubtaskState.FORCING_REPORT,
+            SubtaskState.REPORTED,
+            SubtaskState.FORCING_RESULT_TRANSFER,
+            SubtaskState.RESULT_UPLOADED,
+            SubtaskState.FORCING_ACCEPTANCE,
+            SubtaskState.REJECTED,
+            SubtaskState.ACCEPTED,
+            SubtaskState.FAILED,
+        },
+        'report_computed_task': {
+            SubtaskState.FORCING_REPORT,
+            SubtaskState.REPORTED,
+            SubtaskState.FORCING_RESULT_TRANSFER,
+            SubtaskState.RESULT_UPLOADED,
+            SubtaskState.FAILED,
+        },
+        'ack_report_computed_task': {},
+        'reject_report_computed_task': {},
+        'subtask_results_accepted': {},
+        'subtask_results_rejected': {
+            SubtaskState.REJECTED,
+        },
+    }
```

Comment:
Please remember to always add assertions when you declare a complex constant like this. This is a good way to keep them all in sync.

``` python
assert set(SUBTASK_STORED_MESSAGES_TO_CORRESPONDING_GOLEM_MESSAGE) == set(REQUIRED_RELATED_MESSAGES_IN_STATES)
assert {
    state
    for states in REQUIRED_RELATED_MESSAGES_IN_STATES.values()
    for state in states
}.issubset(set(SubtaskState))
```

Also, it's the second time we have an assert that needs to combine all lists from dict values into a single set so might be a good idea to create a helper function and shorten it too.
``` python
assert flattened_value_set(REQUIRED_RELATED_MESSAGES_IN_STATES).issubset(set(SubtaskState))
```

Reply:
done :+1: Ok I will try to pay more attention to this.